process/datadog_checks/process/process.py

```diff
@@ -109,63 +165,76 @@ def find_pids(self, name, search_string, exact_match, ignore_ad=True):
 
         refresh_ad_cache = self.should_refresh_ad_cache(name)
 
-        encountered_process_names = set()
         matching_pids = set()
 
-        for proc in psutil.process_iter():
-            # Skip access denied processes
-            if not refresh_ad_cache and proc.pid in self.ad_cache:
-                continue
+        # Acquire the write lock to check whether to refresh because we're
+        # going to keep it to do the refresh, AND, we don't want multiple
+        # threads getting a `yes` result at once
+        ProcessCheck.process_list_lock.acquire_write()
+        if self.should_refresh_proclist():
+            self.log.debug("Refreshing process list")
+            ProcessCheck.process_list = [proc for proc in psutil.process_iter(attrs=['pid', 'name'])]
+            ProcessCheck.last_proc_list_ts = time.time()
+            self.log.debug("Set last ts to %s", ProcessCheck.last_proc_list_ts)
+        else:
+            self.log.debug("Using process list cache")
+        ProcessCheck.process_list_lock.release_write()
```

Comment:
Need to be in a finally. It would be nice to have a context manager so that you can use `with` and not care about that. As you have 2 operations, maybe:

```
with self.process_list_cache.write_lock:
    ...

with self.process_list_cache.read_lock:
```